core/embed/lib/error_handling.h

```diff
@@ -20,6 +20,151 @@
+  TS_ERROR_ARG = TS_BUILD(4),      // Invalid argument
+  TS_ERROR_IO = TS_BUILD(5),       // I/O error
+} ts_t;
+
+// Check status code consistency and returns its value.
+// If invalid status code is detected, it will call `__fatal_error()`.
+#define _ts_checked(status)                                       \
+  ({                                                              \
+    ts_t _checked_status = (status);                              \
+    if ((_checked_status & 0xFFFF) != (~_checked_status >> 16)) { \
+      __fatal_error("ts_check() error", __FILE_NAME__, __LINE__); \
+    }                                                             \
+    _checked_status;                                              \
+  })
+
+// Returns `true` if status code is `TS_OK`
+#define ts_ok(status) (_ts_checked(status) == TS_OK)
+
+// Returns `true` if status code is NOT `TS_OK`
+#define ts_error(status) (_ts_checked(status) != TS_OK)
+
+// ----------------------------------------------------
+// verify_init(), verify_status() and verify_xxx() macros define
+// a simple error handling mechanism
+//
+// Example:
+//
+// ts_t my_function(int arg) {
+//   // initialize verify mechanism
+//   verify_init();
+//
+//   // check arguments
+//   verify_arg(arg > 0);
+//
+//   ts_t status;
+//
+//   // verify success
+//   status = some_function();
+//   verify_ok(status);
+//
+//   // verify condition
+//   verify(another_function() != 0, TS_ERROR_IO);
+//
+//  error:
+//
+//   // clean up code comes here
+//
+//   return verify_status();
+// }
+
+// Declares a status variable and initializes it to `TS_OK`.
+// This variable is used to store the status
+#define verify_init() __attribute__((unused)) ts_t __status = TS_OK;
+
+// Returns the current status.
+#define verify_status() (__status)
+
+// Jumps to `error` label if status is not `TS_OK`.
+#define verify_ok(status)    \
+  do {                       \
+    ts_t _status = status;   \
+    if (ts_error(_status)) { \
+      __status = _status;    \
+      goto error;            \
+    }                        \
+  } while (0)
+
+// Jumps to `error` label if the condition is not `true`.
+#define verify(cond, status) \
+  do {                       \
+    if (!(cond)) {           \
+      __status = status;     \
+      goto error;            \
+    }                        \
+  } while (0)
+
+// Jumps to `error` label if the condition is not `true`.
+// Sets the status to `TS_ERROR_ARG`.
+#define verify_arg(cond)       \
+  do {                         \
+    if (!(cond)) {             \
+      __status = TS_ERROR_ARG; \
+      goto error;              \
+    }                          \
+  } while (0)
+
+// Jumps to `error` label if the condition is not `sectrue`.
+#define verify_sec(seccond, status) \
+  do {                              \
+    if ((seccond) != sectrue) {     \
+      __status = status;            \
+      goto error;                   \
+    }                               \
+  } while (0)
```

Comment:
Looking at the macros defined here, it looks like we are not getting rid of secbool. Our code currently uses `bool`, `secbool`, a custom enum (e.g. optiga code) or a custom set of integers (e.g. trezor-crypto) to indicate the return status. The most familiar, compact and easy to use is `bool`. A more secure variant is `secbool`, which should have better resilience against fault injection, but it can be cumbersome to use. The enums are also fine, but they lack type safety. Finally, the custom integers are an abomination. I think that we should be aiming to narrow down and consolidate these different approaches. That could mean for example making secbool less cumbersome to use and possibly building on top of it, but not building yet another approach alongside it.

Reply:
I've been thinking about it too. However, I still don't know how to reasonably connect it. Using two different types to represent boolean values isn't ideal. 

Maybe after the this change cf075b5370487e7a88db7ad3295abd6196c1da4d, 'secfalse' is more like `TS_OK` and `sectrue` is more like TS_BUILD(0xAAAA)... I will think about it.